pkg/platform/runtime/artifact/buildPlan.go

```diff
@@ -0,0 +1,217 @@
+		return
+	}
+
+	var deps []strfmt.UUID
+	for _, depID := range artifact.RuntimeDependencies {
+		deps = append(deps, strfmt.UUID(depID))
+		deps = append(deps, BuildRuntimeDependencies(depID, lookup, deps)...)
+		buildMap(depID, lookup, result)
+	}
+
+	var uniqueDeps []strfmt.UUID
+	for _, dep := range deps {
+		if !funk.Contains(uniqueDeps, dep) {
+			uniqueDeps = append(uniqueDeps, dep)
+		}
+	}
+
+	info, err := GetSourceInfo(artifact.GeneratedBy, lookup)
+	if err != nil {
+		logging.Error("Could not resolve source information: %v", err)
+		return
+	}
+
+	result[strfmt.UUID(artifact.TargetID)] = ArtifactBuildPlan{
+		ArtifactID:       strfmt.UUID(artifact.TargetID),
+		Name:             info.Name,
+		Namespace:        info.Namespace,
+		Version:          &info.Version,
+		RequestedByOrder: true,
+		GeneratedBy:      artifact.GeneratedBy,
+		Dependencies:     uniqueDeps,
+	}
+}
+
+type SourceInfo struct {
+	Name      string
+	Namespace string
+	Version   string
+}
+
+func GetSourceInfo(sourceID string, lookup map[string]interface{}) (SourceInfo, error) {
+	source, ok := lookup[sourceID].(*model.Source)
+	if ok {
+		return SourceInfo{source.Name, source.Namespace, source.Version}, nil
+	}
+
+	step, ok := lookup[sourceID].(*model.Step)
+	if !ok {
+		return SourceInfo{}, locale.NewError("err_source_name_step", "Could not find step with generatedBy id {{.V0}}", sourceID)
+	}
+
+	for _, input := range step.Inputs {
+		if input.Tag != model.TagSource {
+			continue
+		}
+		for _, id := range input.TargetIDs {
+			source, ok := lookup[id].(*model.Source)
+			if !ok {
+				return SourceInfo{}, locale.NewError("err_source_name_source", "Could not find source with target id {{.V0}}", id)
+			}
+			return SourceInfo{source.Name, source.Namespace, source.Version}, nil
+		}
+	}
+	return SourceInfo{}, locale.NewError("err_resolve_artifact_name", "Could not resolve artifact name")
+}
+
+func BuildRuntimeDependencies(depdendencyID string, lookup map[string]interface{}, result []strfmt.UUID) []strfmt.UUID {
+	artifact, ok := lookup[depdendencyID].(*model.Artifact)
+	if !ok {
+		logging.Error("Incorrect target type for id %s", depdendencyID)
+	}
+
+	for _, depID := range artifact.RuntimeDependencies {
+		result = append(result, strfmt.UUID(depID))
+		BuildRuntimeDependencies(depID, lookup, result)
+	}
+
+	return result
+}
+
+// RecursiveDependenciesFor computes the recursive dependencies for an ArtifactID a using artifacts as a lookup table
+func RecursiveDependenciesFor(a ArtifactID, artifacts ArtifactBuildPlanMap) []ArtifactID {
+	allDeps := make(map[ArtifactID]struct{})
+	artf, ok := artifacts[a]
+	if !ok {
+		return nil
+	}
+	toCheck := artf.Dependencies
+
+	for len(toCheck) > 0 {
+		var newToCheck []ArtifactID
+		for _, a := range toCheck {
+			if _, ok := allDeps[a]; ok {
+				continue
+			}
+			artf, ok := artifacts[a]
+			if !ok {
+				continue
+			}
+			newToCheck = append(newToCheck, artf.Dependencies...)
+			allDeps[a] = struct{}{}
```

Comment:
This map structure was chosen for cheap future proofing since we do not know how large these build plans will get. The `map[string]struct{}` traditionally outperforms the `map[string]bool` in both time and memory. This could have changed with newer versions of Go.